hrms/payroll/doctype/salary_slip/salary_slip.py

```diff
@@ -610,21 +639,255 @@ def set_net_pay(self):
+			)
+
+	def compute_ctc(self):
+		if hasattr(self, "previous_taxable_earnings"):
+			return (
+				self.previous_taxable_earnings
+				+ self.current_structured_taxable_earnings
+				+ self.future_structured_taxable_earnings
+				+ self.current_additional_earnings
+				+ self.other_incomes
+				+ self.unclaimed_taxable_benefits
+				+ self.non_taxable_earnings
+			)
+
+		else:
+			return 0.0
+
+	def compute_non_taxable_earnings(self):
+		non_taxable_earnings = 0.0
+		non_taxable_additional_salary = 0.0
+
+		prev_period_non_taxable_earnings = 0.0
+		current_period_non_taxable_earnings = 0.0
+		future_period_non_taxable_earnings = 0.0
+
+		# Previous period non taxable earnings
+		prev_period_non_taxable_earnings = self.get_salary_slip_details(
+			self.payroll_period.start_date, self.start_date, parentfield="earnings", is_tax_applicable=0
+		)
+
+		non_taxable_additional_salary = self.get_salary_slip_details(
+			self.payroll_period.start_date,
+			self.start_date,
+			parentfield="earnings",
+			is_tax_applicable=0,
+			field_to_select="additional_amount",
+		)
+
+		# Current period non taxable earnings
+		for earning in self.earnings:
+			if not earning.is_tax_applicable:
+				if not earning.additional_amount:
+					current_period_non_taxable_earnings += earning.amount
+
+				if earning.additional_amount:
+					non_taxable_additional_salary += earning.additional_amount
+
+					# Future recurring additional salary
+					if earning.additional_salary and earning.is_recurring_additional_salary:
+						non_taxable_additional_salary += self.get_future_recurring_additional_amount(
+							earning.additional_salary, earning.additional_amount
+						)
+
+		# Future period non taxable earnings
+		future_period_non_taxable_earnings += current_period_non_taxable_earnings * (
+			math.ceil(self.remaining_sub_periods) - 1
+		)
+
+		non_taxable_earnings = (
+			prev_period_non_taxable_earnings
+			+ current_period_non_taxable_earnings
+			+ future_period_non_taxable_earnings
+			+ non_taxable_additional_salary
+		)
+
+		return non_taxable_earnings
+
+	def compute_annual_deductions_before_tax_calculation(self):
+		prev_period_exempted_amount = 0
+		current_period_exempted_amount = 0
+		future_period_exempted_amount = 0
+
+		# Previous period exempted amount
+		prev_period_exempted_amount = self.get_salary_slip_details(
+			self.payroll_period.start_date,
+			self.start_date,
+			parentfield="deductions",
+			exempted_from_income_tax=1,
+		)
+
+		# Current period exempted amount
+		for d in self.get("deductions"):
+			if d.exempted_from_income_tax:
+				current_period_exempted_amount += d.amount
+
+		# Future period exempted amount
+		for deduction in self._salary_structure_doc.get("deductions"):
+			if deduction.exempted_from_income_tax:
+				if deduction.amount_based_on_formula:
+					for sub_period in range(1, math.ceil(self.remaining_sub_periods)):
+						future_period_exempted_amount += self.get_amount_from_formula(deduction, sub_period)
+					pass
```

Comment:
Any reason to pass here? It's followed by an else statement anyway so won't execute the iteration further